CalculatePolynomialBackground

soininen opened the pull request:
This PR introduces a new `CalculatePolynomialBackground` algorithm. As the name states, the algorithm fits a polynomial of given degree to histograms in a workspace and returns the evaluated values.

As a side effect, `Mantid::CurveFitting::Jacobian` was moved to `Mantid::API` to provide access to a Jacobian implementation. The class was further renamed to `BasicJacobian` to avoid name clash with the abstract `Mantid::API::Jacobian`.

**To test:**

Try surprising it with confusing input. Use your favorite datasets or play with the following script as a starting point:
```python
import numpy

ws = CreateWorkspace(
    DataX=numpy.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]),
    DataY=numpy.array([1, 2, 1, 2, 1, 2, 1, 2, 1]),
)

polybkg = CalculatePolynomialBackground(ws)
```

Resolves #20404.

**Release Notes** 

Release notes were updated accordingly.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

samueljackson92 commented on Framework/Algorithms/src/CalculatePolynomialBackground.cpp:
```diff
@@ -0,0 +1,279 @@
+//----------------------------------------------------------------------------------------------
+/** Execute the algorithm.
+ */
+void CalculatePolynomialBackground::exec() {
```
This `exec` method does too much. From a brief glance I can see that there could be separate methods for:
 - Finding the included range
 - Running the fit
 - Pulling the parameters & errors out for the fit workspace
 - Evaluating the background function and filling the Y/E values

samueljackson92 commented on Framework/Algorithms/test/CalculatePolynomialBackgroundTest.h:
```diff
@@ -0,0 +1,152 @@
+    a->setRethrows(true);
+    return a;
+  }
+};
```
Could we also add a performance test?

soininen replied:
Great idea!

samueljackson92 commented on Framework/CurveFitting/test/Functions/ProductQuadraticExpTest.h:
```diff
@@ -199,7 +199,7 @@ to check that the results are equal.
     FunctionDomain1DVector domain(xValues);
 
-    Mantid::CurveFitting::Jacobian jacobian(nResults, 5);
+    Mantid::API::BasicJacobian jacobian(nResults, 5);
```
Kind of feels like this abstract base class should be changed to be called `IJacobian` or something instead and then call this just `Jacobian`, but perhaps that touches too much code to change?

soininen commented:
I've removed the error calculations from the algorithm for the moment and reverted the Jacobian move/rename. It seems to be easy to construct cases where the errors are way overestimated the way they are calculated now.

NickDraper commented:
@samueljackson92 Could you take another look at this one please

samueljackson92 commented:
@soininen With the script I posted above I now get a very bad fit. This is much worse than the original fit with large errors.

![screen shot 2017-09-26 at 16 35 11](https://user-images.githubusercontent.com/2487781/30869301-af5fea60-a2d8-11e7-9866-670d37bc3965.png)

soininen commented:
@samueljackson92 Thanks for noticing a bug! A classical off-by-one error had slipped in.

samueljackson92 commented:
@soininen  This seems to be giving sensible results now. Looks good to me. 👍
